examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,417 @@
+the Triplet Loss using the three embeddings produced by the Siamese network.
+
+Here is the definition of Triplet Loss implemented on this example:
+`L(A,P,N) = max(||f(A)-f(P)||**2 - ||f(A)-f(N)||**2 + alpha, 0)`
+"""
+
+"""
+Let's create a `Mean` metric instance to track the loss of the training process.
+"""
+
+loss_tracker = metrics.Mean(name="loss")
+
+
+class SiameseModel(Model):
+    """
+    Model implementing a custom training loop to compute the Triplet Loss
+    using the three embeddings produced by the Siamese network.
+
+    Here is the definition of Triplet Loss:
+        L(A,P,N) = max(||f(A)-f(P)||**2 - ||f(A)-f(N)||**2 + alpha, 0)
+
+    """
+
+    def __init__(self, siamese_network, alpha=0.5):
+        super(SiameseModel, self).__init__()
+        self.siamese_network = siamese_network
+        self.alpha = alpha
+
+    def call(self, inputs):
+        self.siamese_network(inputs)
+
+    def train_step(self, data):
+        # GradientTape is a context manager that records every operation that you do inside.
+        # We are using it here to compute the loss so we can get the gradients and apply
+        # them using the optimizer specified in `compile()`.
+        with tf.GradientTape() as tape:
+            anchor, positive, negative = data
+
+            # The output of the network is a tuple containing the distances
+            # between the anchor and the positive example, and the anchor and
+            # the negative example.
+            ap_distance, an_distance = self.siamese_network(
+                (anchor, positive, negative)
+            )
+
+            # Computing the Triplet Loss by subtracting both distances and
+            # making sure we don't get a negative value.
+            loss = ap_distance - an_distance
+            loss = tf.maximum(loss + self.alpha, 0.0)
+
+        # Let's get the gradients (loss with respect to trainable weights)
+        gradients = tape.gradient(loss, self.siamese_network.trainable_weights)
+
+        # Applying the gradients on the model using the specified optimizer
+        self.optimizer.apply_gradients(
+            zip(gradients, self.siamese_network.trainable_weights)
+        )
+
+        # Let's update and return the loss metric.
+        loss_tracker.update_state(loss)
+        return {"loss": loss_tracker.result()}
+
+    @property
+    def metrics(self):
+        # We need to list our metric here so the `reset_states()` can be
+        # called automatically.
+        return [loss_tracker]
+
+
+"""
+# Training
+
+We are now ready to train our model.
+"""
+
+siamese_model = SiameseModel(siamese_network)
+siamese_model.compile(optimizer=optimizers.Adam(0.0001))
+siamese_model.fit(dataset, epochs=20)
+
+"""
+# Looking at what the network learned
+
+At this point, we can check how the network learned to separate the embeddings
+depending on whether they belong to similar images.
+
+We can use Cosine Similarity to measure the similarity between embeddings.
```

Comment:
Suggestion: provide a link, use sentence-case

```suggestion
We can use [cosine similarity](https://en.wikipedia.org/wiki/Cosine_similarity) to measure the
similarity between embeddings.
```